// twitter.js/dist/esm/managers/FilteredStreamRuleManager.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FilteredStreamRuleManager_instances, _FilteredStreamRuleManager_deleteRules, _FilteredStreamRuleManager_fetchSingleRule, _FilteredStreamRuleManager_fetchMultipleRules;
import { Collection } from '../util';
import { BaseManager } from './BaseManager';
import { CustomError, CustomTypeError } from '../errors';
import { RequestData, FilteredStreamRule } from '../structures';
export class FilteredStreamRuleManager extends BaseManager {
    /**
     * @param client The logged in {@link Client} instance
     */
    constructor(client) {
        super(client, FilteredStreamRule);
        _FilteredStreamRuleManager_instances.add(this);
    }
    /**
     * Fetches one or multiple rules that are currently active.
     * @param options The options for fetching rules
     * @returns A {@link FilteredStreamRule} or a {@link Collection} of them
     */
    async fetch(options) {
        if (typeof options === 'undefined') {
            return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_fetchMultipleRules).call(this);
        }
        else if ('rule' in options) {
            const ruleId = this.resolveId(options.rule);
            if (!ruleId)
                throw new CustomError('RULE_RESOLVE_ID', 'fetch');
            return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_fetchSingleRule).call(this, ruleId, options);
        }
        else if ('rules' in options) {
            if (typeof options.rules === 'undefined') {
                return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_fetchMultipleRules).call(this, options.rules, options);
            }
            if (!Array.isArray(options.rules)) {
                throw new CustomTypeError('INVALID_TYPE', 'rules', 'array or undefined', true);
            }
            const ruleIds = options.rules.map(rule => {
                const ruleId = this.resolveId(rule);
                if (!ruleId)
                    throw new CustomError('RULE_RESOLVE_ID', 'fetch');
                return ruleId;
            });
            return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_fetchMultipleRules).call(this, ruleIds, options);
        }
        return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_fetchMultipleRules).call(this);
    }
    /**
     * Creates one or multiple rules for the filtered stream.
     * @param data The data for creating rules
     * @returns A {@link Collection} of {@link FilteredStreamRule} objects
     */
    async create(data) {
        const rules = Array.isArray(data) ? data : [data];
        const body = {
            add: rules,
        };
        const requestData = new RequestData({ body });
        const res = await this.client._api.tweets.search.stream.rules.post(requestData);
        return res.data.reduce((createdRules, rawRule) => {
            const rule = this._add(rawRule.id, rawRule);
            return createdRules.set(rule.id, rule);
        }, new Collection());
    }
    /**
     * Deletes one or multiple rules for the filtered stream using their ids.
     * @param ruleId The id or ids of the rules to delete
     */
    async deleteById(ruleId) {
        const ids = Array.isArray(ruleId) ? ruleId : [ruleId];
        const body = {
            delete: {
                ids,
            },
        };
        return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_deleteRules).call(this, body);
    }
    /**
     * Deletes one or multiple rules for the filtered stream using their values.
     * @param ruleValue The value or values of the rules to delete
     */
    async deleteByValue(ruleValue) {
        const values = Array.isArray(ruleValue) ? ruleValue : [ruleValue];
        const body = {
            delete: {
                values,
            },
        };
        return __classPrivateFieldGet(this, _FilteredStreamRuleManager_instances, "m", _FilteredStreamRuleManager_deleteRules).call(this, body);
    }
}
_FilteredStreamRuleManager_instances = new WeakSet(), _FilteredStreamRuleManager_deleteRules = 
// #### 🚧 PRIVATE METHODS 🚧 ####
async function _FilteredStreamRuleManager_deleteRules(body) {
    const requestData = new RequestData({ body });
    const res = await this.client._api.tweets.search.stream.rules.post(requestData);
    return res;
}, _FilteredStreamRuleManager_fetchSingleRule = async function _FilteredStreamRuleManager_fetchSingleRule(ruleId, options) {
    if (!options.skipCacheCheck) {
        const cachedRule = this.cache.get(ruleId);
        if (cachedRule)
            return cachedRule;
    }
    const query = {
        ids: [ruleId],
    };
    const requestData = new RequestData({ query });
    const res = await this.client._api.tweets.search.stream.rules.get(requestData);
    const rawRule = res.data?.[0];
    if (!rawRule)
        throw new CustomError('RULE_NOT_FOUND');
    return this._add(rawRule.id, rawRule, options.cacheAfterFetching);
}, _FilteredStreamRuleManager_fetchMultipleRules = async function _FilteredStreamRuleManager_fetchMultipleRules(ruleIds, options) {
    const fetchedRules = new Collection();
    const query = {
        ids: ruleIds,
    };
    const requestData = new RequestData({ query });
    const res = await this.client._api.tweets.search.stream.rules.get(requestData);
    const rawRules = res.data;
    if (!rawRules?.length)
        return fetchedRules;
    for (const rawRule of rawRules) {
        const rule = this._add(rawRule.id, rawRule, options?.cacheAfterFetching);
        fetchedRules.set(rule.id, rule);
    }
    return fetchedRules;
};
