// twitter.js/dist/esm/books/FollowingsBook.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _FollowingsBook_instances, _FollowingsBook_nextToken, _FollowingsBook_previousToken, _FollowingsBook_hasMadeInitialRequest, _FollowingsBook_fetchPages;
import { Collection } from '../util';
import { BaseBook } from './BaseBook';
import { CustomError } from '../errors';
import { RequestData } from '../structures';
/**
 * A class for fetching users followed by a twitter user
 */
export class FollowingsBook extends BaseBook {
    /**
     * @param client The logged in {@link Client} instance
     * @param options The options to initialize the followings book with
     */
    constructor(client, options) {
        super(client);
        _FollowingsBook_instances.add(this);
        /**
         * The token for fetching next page
         */
        _FollowingsBook_nextToken.set(this, void 0);
        /**
         * The token for fetching previous page
         */
        _FollowingsBook_previousToken.set(this, void 0);
        /**
         * Whether an initial request for fetching the first page has already been made
         *
         * **Note**: Use this to not throw `PAGINATED_RESPONSE_TAIL_REACHED` error for initial page request in {@link FollowingsBook.fetchNextPage}
         */
        _FollowingsBook_hasMadeInitialRequest.set(this, void 0);
        this.hasMore = true;
        this.userId = options.userId;
        this.maxResultsPerPage = options.maxResultsPerPage ?? null;
    }
    /**
     * Fetches the next page of the book if there is one.
     * @returns A {@link Collection} of {@link User} objects that the owner of this book is following
     */
    async fetchNextPage() {
        if (!__classPrivateFieldGet(this, _FollowingsBook_hasMadeInitialRequest, "f")) {
            __classPrivateFieldSet(this, _FollowingsBook_hasMadeInitialRequest, true, "f");
            return __classPrivateFieldGet(this, _FollowingsBook_instances, "m", _FollowingsBook_fetchPages).call(this);
        }
        if (!__classPrivateFieldGet(this, _FollowingsBook_nextToken, "f"))
            throw new CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _FollowingsBook_instances, "m", _FollowingsBook_fetchPages).call(this, __classPrivateFieldGet(this, _FollowingsBook_nextToken, "f"));
    }
    /**
     * Fetches the previous page of the book if there is one.
     * @returns A {@link Collection} of {@link User} objects that the owner of this book is following
     */
    async fetchPreviousPage() {
        if (!__classPrivateFieldGet(this, _FollowingsBook_previousToken, "f"))
            throw new CustomError('PAGINATED_RESPONSE_HEAD_REACHED');
        return __classPrivateFieldGet(this, _FollowingsBook_instances, "m", _FollowingsBook_fetchPages).call(this, __classPrivateFieldGet(this, _FollowingsBook_previousToken, "f"));
    }
}
_FollowingsBook_nextToken = new WeakMap(), _FollowingsBook_previousToken = new WeakMap(), _FollowingsBook_hasMadeInitialRequest = new WeakMap(), _FollowingsBook_instances = new WeakSet(), _FollowingsBook_fetchPages = 
// #### 🚧 PRIVATE METHODS 🚧 ####
async function _FollowingsBook_fetchPages(token) {
    const followingsCollection = new Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.userExpansions,
        'tweet.fields': queryParameters?.tweetFields,
        'user.fields': queryParameters?.userFields,
        pagination_token: token,
    };
    if (this.maxResultsPerPage)
        query.max_results = this.maxResultsPerPage;
    const requestData = new RequestData({ query });
    const data = await this.client._api.users(this.userId).following.get(requestData);
    __classPrivateFieldSet(this, _FollowingsBook_nextToken, data.meta.next_token, "f");
    __classPrivateFieldSet(this, _FollowingsBook_previousToken, data.meta.previous_token, "f");
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return followingsCollection;
    const rawUsers = data.data;
    const rawIncludes = data.includes;
    for (const rawUser of rawUsers) {
        const user = this.client.users._add(rawUser.id, { data: rawUser, includes: rawIncludes });
        followingsCollection.set(user.id, user);
    }
    return followingsCollection;
};
