// twitter.js/dist/esm/books/LikedTweetsBook.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _LikedTweetsBook_instances, _LikedTweetsBook_nextToken, _LikedTweetsBook_previousToken, _LikedTweetsBook_hasMadeInitialRequest, _LikedTweetsBook_fetchPages;
import { Collection } from '../util';
import { BaseBook } from './BaseBook';
import { CustomError } from '../errors';
import { RequestData } from '../structures';
/**
 * A class for fetching tweets liked by a twitter user
 */
export class LikedTweetsBook extends BaseBook {
    /**
     * @param client The logged in {@link Client} instance
     * @param options The options to initialize the liked tweets book with
     */
    constructor(client, options) {
        super(client);
        _LikedTweetsBook_instances.add(this);
        /**
         * The token for fetching next page
         */
        _LikedTweetsBook_nextToken.set(this, void 0);
        /**
         * The token for fetching previous page
         */
        _LikedTweetsBook_previousToken.set(this, void 0);
        /**
         * Whether an initial request for fetching the first page has already been made
         *
         * **Note**: Use this to not throw `PAGINATED_RESPONSE_TAIL_REACHED` error for initial page request in {@link FollowersBook.fetchNextPage}
         */
        _LikedTweetsBook_hasMadeInitialRequest.set(this, void 0);
        this.hasMore = true;
        this.userId = options.userId;
        this.maxResultsPerPage = options.maxResultsPerPage ?? null;
    }
    /**
     * Fetches the next page of the book if there is one.
     * @returns A {@link Collection} of {@link Tweet} objects liked by the owner of this book
     */
    async fetchNextPage() {
        if (!__classPrivateFieldGet(this, _LikedTweetsBook_hasMadeInitialRequest, "f")) {
            __classPrivateFieldSet(this, _LikedTweetsBook_hasMadeInitialRequest, true, "f");
            return __classPrivateFieldGet(this, _LikedTweetsBook_instances, "m", _LikedTweetsBook_fetchPages).call(this);
        }
        if (!__classPrivateFieldGet(this, _LikedTweetsBook_nextToken, "f"))
            throw new CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _LikedTweetsBook_instances, "m", _LikedTweetsBook_fetchPages).call(this, __classPrivateFieldGet(this, _LikedTweetsBook_nextToken, "f"));
    }
    /**
     * Fetches the previous page of the book if there is one.
     * @returns A {@link Collection} of {@link Tweet} objects liked by the owner of this book
     */
    async fetchPreviousPage() {
        if (!__classPrivateFieldGet(this, _LikedTweetsBook_previousToken, "f"))
            throw new CustomError('PAGINATED_RESPONSE_HEAD_REACHED');
        return __classPrivateFieldGet(this, _LikedTweetsBook_instances, "m", _LikedTweetsBook_fetchPages).call(this, __classPrivateFieldGet(this, _LikedTweetsBook_previousToken, "f"));
    }
}
_LikedTweetsBook_nextToken = new WeakMap(), _LikedTweetsBook_previousToken = new WeakMap(), _LikedTweetsBook_hasMadeInitialRequest = new WeakMap(), _LikedTweetsBook_instances = new WeakSet(), _LikedTweetsBook_fetchPages = 
// #### 🚧 PRIVATE METHODS 🚧 ####
async function _LikedTweetsBook_fetchPages(token) {
    const likedTweetsCollection = new Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.tweetExpansions,
        'media.fields': queryParameters?.mediaFields,
        'place.fields': queryParameters?.placeFields,
        'poll.fields': queryParameters?.pollFields,
        'tweet.fields': queryParameters?.tweetFields,
        'user.fields': queryParameters?.userFields,
        pagination_token: token,
    };
    if (this.maxResultsPerPage)
        query.max_results = this.maxResultsPerPage;
    const requestData = new RequestData({ query });
    const data = await this.client._api.users(this.userId).liked_tweets.get(requestData);
    __classPrivateFieldSet(this, _LikedTweetsBook_nextToken, data.meta.next_token, "f");
    __classPrivateFieldSet(this, _LikedTweetsBook_previousToken, data.meta.previous_token, "f");
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return likedTweetsCollection;
    const rawTweets = data.data;
    const rawIncludes = data.includes;
    for (const rawTweet of rawTweets) {
        const tweet = this.client.tweets._add(rawTweet.id, { data: rawTweet, includes: rawIncludes });
        likedTweetsCollection.set(tweet.id, tweet);
    }
    return likedTweetsCollection;
};
