// twitter.js/dist/esm/books/SearchTweetsBook.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _SearchTweetsBook_instances, _SearchTweetsBook_nextToken, _SearchTweetsBook_hasMadeInitialRequest, _SearchTweetsBook_fetchPages;
import { Collection } from '../util';
import { BaseBook } from './BaseBook';
import { CustomError } from '../errors';
import { RequestData } from '../structures';
/**
 * A class for fetching tweets using search query
 */
export class SearchTweetsBook extends BaseBook {
    /**
     * @param client The logged in {@link Client} instance
     * @param options The options to initialize the search tweets book with
     */
    constructor(client, options) {
        super(client);
        _SearchTweetsBook_instances.add(this);
        /**
         * The token for fetching next page
         */
        _SearchTweetsBook_nextToken.set(this, void 0);
        /**
         * Whether an initial request for fetching the first page has already been made
         *
         * **Note**: Use this to not throw `PAGINATED_RESPONSE_TAIL_REACHED` error for initial page request in {@link FollowingsBook.fetchNextPage}
         */
        _SearchTweetsBook_hasMadeInitialRequest.set(this, void 0);
        this.hasMore = true;
        this.query = options.query;
        this.afterTweetId = options.afterTweetId ?? null;
        this.beforeTweetId = options.beforeTweetId ?? null;
        this.afterTimestamp = options.afterTimestamp ?? null;
        this.beforeTimestamp = options.beforeTimestamp ?? null;
        this.maxResultsPerPage = options.maxResultsPerPage ?? null;
    }
    /**
     * Fetches the next page of the book if there is one.
     * @returns A {@link Collection} of {@link Tweet} objects matching the search query
     */
    async fetchNextPage() {
        if (!__classPrivateFieldGet(this, _SearchTweetsBook_hasMadeInitialRequest, "f")) {
            __classPrivateFieldSet(this, _SearchTweetsBook_hasMadeInitialRequest, true, "f");
            return __classPrivateFieldGet(this, _SearchTweetsBook_instances, "m", _SearchTweetsBook_fetchPages).call(this);
        }
        if (!__classPrivateFieldGet(this, _SearchTweetsBook_nextToken, "f"))
            throw new CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _SearchTweetsBook_instances, "m", _SearchTweetsBook_fetchPages).call(this, __classPrivateFieldGet(this, _SearchTweetsBook_nextToken, "f"));
    }
}
_SearchTweetsBook_nextToken = new WeakMap(), _SearchTweetsBook_hasMadeInitialRequest = new WeakMap(), _SearchTweetsBook_instances = new WeakSet(), _SearchTweetsBook_fetchPages = 
// #### 🚧 PRIVATE METHODS 🚧 ####
async function _SearchTweetsBook_fetchPages(token) {
    const tweetsCollection = new Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.tweetExpansions,
        'tweet.fields': queryParameters?.tweetFields,
        'user.fields': queryParameters?.userFields,
        'media.fields': queryParameters?.mediaFields,
        'place.fields': queryParameters?.placeFields,
        'poll.fields': queryParameters?.pollFields,
        query: this.query,
        next_token: token,
    };
    if (this.afterTweetId)
        query.since_id = this.afterTweetId;
    if (this.beforeTweetId)
        query.until_id = this.beforeTweetId;
    if (this.maxResultsPerPage)
        query.max_results = this.maxResultsPerPage;
    if (this.afterTimestamp)
        query.start_time = new Date(this.afterTimestamp).toISOString();
    if (this.beforeTimestamp)
        query.end_time = new Date(this.beforeTimestamp).toISOString();
    const requestData = new RequestData({ query });
    const data = await this.client._api.tweets.search.recent.get(requestData);
    __classPrivateFieldSet(this, _SearchTweetsBook_nextToken, data.meta.next_token, "f");
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return tweetsCollection;
    const rawTweets = data.data;
    const rawIncludes = data.includes;
    for (const rawTweet of rawTweets) {
        const tweet = this.client.tweets._add(rawTweet.id, { data: rawTweet, includes: rawIncludes });
        tweetsCollection.set(tweet.id, tweet);
    }
    return tweetsCollection;
};
