// twitter.js/dist/esm/books/TweetsCountBook.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _TweetsCountBook_instances, _TweetsCountBook_nextToken, _TweetsCountBook_hasMadeInitialRequest, _TweetsCountBook_fetchPages;
import { BaseBook } from './BaseBook';
import { CustomError } from '../errors';
import { RequestData, TweetCountBucket } from '../structures';
/**
 * A class for fetching number of tweets matching a search query
 */
export class TweetsCountBook extends BaseBook {
    /**
     * @param client The logged in {@link Client} instance
     * @param options The options to initialize the count tweets book with
     */
    constructor(client, options) {
        super(client);
        _TweetsCountBook_instances.add(this);
        /**
         * The token for fetching next page
         */
        _TweetsCountBook_nextToken.set(this, void 0);
        /**
         * Whether an initial request for fetching the first page has already been made
         *
         * **Note**: Use this to **not** throw `PAGINATED_RESPONSE_TAIL_REACHED` error for initial page request in {@link BlocksBook.fetchNextPage}
         */
        _TweetsCountBook_hasMadeInitialRequest.set(this, void 0);
        this.hasMore = true;
        this.query = options.query;
        this.granularity = options.granularity ?? null;
        this.afterTweetId = options.afterTweetId ?? null;
        this.beforeTweetId = options.beforeTweetId ?? null;
        this.afterTimestamp = options.afterTimestamp ?? null;
        this.beforeTimestamp = options.beforeTimestamp ?? null;
    }
    /**
     * Fetches the next page of the book if there is one.
     * @returns
     */
    async fetchNextPage() {
        if (!__classPrivateFieldGet(this, _TweetsCountBook_hasMadeInitialRequest, "f")) {
            __classPrivateFieldSet(this, _TweetsCountBook_hasMadeInitialRequest, true, "f");
            return __classPrivateFieldGet(this, _TweetsCountBook_instances, "m", _TweetsCountBook_fetchPages).call(this);
        }
        if (!__classPrivateFieldGet(this, _TweetsCountBook_nextToken, "f"))
            throw new CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _TweetsCountBook_instances, "m", _TweetsCountBook_fetchPages).call(this, __classPrivateFieldGet(this, _TweetsCountBook_nextToken, "f"));
    }
}
_TweetsCountBook_nextToken = new WeakMap(), _TweetsCountBook_hasMadeInitialRequest = new WeakMap(), _TweetsCountBook_instances = new WeakSet(), _TweetsCountBook_fetchPages = 
// #### 🚧 PRIVATE METHODS 🚧 ####
async function _TweetsCountBook_fetchPages(token) {
    const tweetCountBuckets = [];
    const query = {
        query: this.query,
        next_token: token,
    };
    if (this.granularity)
        query.granularity = this.granularity;
    if (this.afterTweetId)
        query.since_id = this.afterTweetId;
    if (this.beforeTweetId)
        query.until_id = this.beforeTweetId;
    if (this.afterTimestamp)
        query.start_time = new Date(this.afterTimestamp).toISOString();
    if (this.beforeTimestamp)
        query.end_time = new Date(this.beforeTimestamp).toISOString();
    const requestData = new RequestData({ query });
    const data = await this.client._api.tweets.counts.recent.get(requestData);
    __classPrivateFieldSet(this, _TweetsCountBook_nextToken, data.meta.next_token, "f");
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.total_tweet_count === 0)
        return tweetCountBuckets;
    const rawBuckets = data.data;
    for (const rawBucket of rawBuckets) {
        const bucket = new TweetCountBucket(rawBucket, this.granularity);
        tweetCountBuckets.push(bucket);
    }
    return tweetCountBuckets;
};
