// twitter.js/dist/cjs/books/BlocksBook.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _BlocksBook_instances, _BlocksBook_nextToken, _BlocksBook_previousToken, _BlocksBook_hasMadeInitialRequest, _BlocksBook_fetchPages;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlocksBook = void 0;
const util_1 = require("../util");
const BaseBook_1 = require("./BaseBook");
const errors_1 = require("../errors");
const structures_1 = require("../structures");
/**
 * A class for fetching users blocked by the authorized user
 */
class BlocksBook extends BaseBook_1.BaseBook {
    /**
     * @param client The logged in {@link Client} instance
     * @param options The options to initialize the blocks book with
     */
    constructor(client, options) {
        super(client);
        _BlocksBook_instances.add(this);
        /**
         * The token for fetching next page
         */
        _BlocksBook_nextToken.set(this, void 0);
        /**
         * The token for fetching previous page
         */
        _BlocksBook_previousToken.set(this, void 0);
        /**
         * Whether an initial request for fetching the first page has already been made
         *
         * **Note**: Use this to not throw `PAGINATED_RESPONSE_TAIL_REACHED` error for initial page request in {@link BlocksBook.fetchNextPage}
         */
        _BlocksBook_hasMadeInitialRequest.set(this, void 0);
        this.userId = options.userId;
        this.maxResultsPerPage = options.maxResultsPerPage ?? null;
        this.hasMore = true;
    }
    /**
     * Fetches the next page of the book if there is one.
     * @returns A {@link Collection} of {@link User} objects that have been blocked by the authorized user
     */
    async fetchNextPage() {
        if (!__classPrivateFieldGet(this, _BlocksBook_hasMadeInitialRequest, "f")) {
            __classPrivateFieldSet(this, _BlocksBook_hasMadeInitialRequest, true, "f");
            return __classPrivateFieldGet(this, _BlocksBook_instances, "m", _BlocksBook_fetchPages).call(this);
        }
        if (!__classPrivateFieldGet(this, _BlocksBook_nextToken, "f"))
            throw new errors_1.CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _BlocksBook_instances, "m", _BlocksBook_fetchPages).call(this, __classPrivateFieldGet(this, _BlocksBook_nextToken, "f"));
    }
    /**
     * Fetches the previous page of the book if there is one.
     * @returns A {@link Collection} of {@link User} objects that have been blocked by the authorized user
     */
    async fetchPreviousPage() {
        if (!__classPrivateFieldGet(this, _BlocksBook_previousToken, "f"))
            throw new errors_1.CustomError('PAGINATED_RESPONSE_HEAD_REACHED');
        return __classPrivateFieldGet(this, _BlocksBook_instances, "m", _BlocksBook_fetchPages).call(this, __classPrivateFieldGet(this, _BlocksBook_previousToken, "f"));
    }
}
exports.BlocksBook = BlocksBook;
_BlocksBook_nextToken = new WeakMap(), _BlocksBook_previousToken = new WeakMap(), _BlocksBook_hasMadeInitialRequest = new WeakMap(), _BlocksBook_instances = new WeakSet(), _BlocksBook_fetchPages = 
// #### 🚧 PRIVATE METHODS 🚧 ####
async function _BlocksBook_fetchPages(token) {
    const blockedUsersCollection = new util_1.Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.userExpansions,
        'tweet.fields': queryParameters?.tweetFields,
        'user.fields': queryParameters?.userFields,
        pagination_token: token,
    };
    if (this.maxResultsPerPage)
        query.max_results = this.maxResultsPerPage;
    const requestData = new structures_1.RequestData({ query, isUserContext: true });
    const data = await this.client._api.users(this.userId).blocking.get(requestData);
    __classPrivateFieldSet(this, _BlocksBook_nextToken, data.meta.next_token, "f");
    __classPrivateFieldSet(this, _BlocksBook_previousToken, data.meta.previous_token, "f");
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return blockedUsersCollection;
    const rawUsers = data.data;
    const rawIncludes = data.includes;
    for (const rawUser of rawUsers) {
        const user = this.client.users._add(rawUser.id, { data: rawUser, includes: rawIncludes }, false);
        blockedUsersCollection.set(user.id, user);
    }
    return blockedUsersCollection;
};
