// twitter.js/dist/esm/structures/Media.js
import { MediaNonPublicMetrics, MediaOrganicMetrics, MediaPromotedMetrics, MediaPublicMetrics } from './misc';
/**
 * The class that represents a media content in a {@link Tweet}
 */
export class Media {
    constructor(client, data) {
        Object.defineProperty(this, 'client', { writable: true, enumerable: false });
        this.client = client;
        this.id = data.media_key;
        this.type = data.type;
        this.duration = data.duration_ms ?? null;
        this.height = data.height ?? null;
        this.nonPublicMetrics = data.non_public_metrics ? new MediaNonPublicMetrics(data.non_public_metrics) : null;
        this.organicMetrics = data.organic_metrics ? new MediaOrganicMetrics(data.organic_metrics) : null;
        this.previewImageURL = data.preview_image_url ?? null;
        this.promotedMetrics = data.promoted_metrics ? new MediaPromotedMetrics(data.promoted_metrics) : null;
        this.publicMetrics = data.public_metrics ? new MediaPublicMetrics(data.public_metrics) : null;
        this.url = data.url ?? null;
        this.width = data.width ?? null;
    }
}
